R/group_apply.R

```diff
@@ -0,0 +1,133 @@
+#' grouping variable.
+#' - \code{\link{tax_range_geo}}: return the geographic range of taxa per
+#' grouping variable.
+#' - \code{\link{tax_check}}: return potential spelling variations of the
+#' same taxon per grouping variable. Note: `verbose` needs to be set to FALSE.
+#'
+#'
+#' @section Developer(s):
+#' Lewis A. Jones & William Gearty
+#' @section Reviewer(s):
+#' Kilian Eichenseer & Bethany Allen
+#' @importFrom stats as.formula
+#' @examples
+#' # Examples
+#' # Get tetrapods data
+#' occdf <- tetrapods
+#' # Remove NA data
+#' occdf <- subset(occdf, !is.na(genus))
+#' # Count number of occurrences from each country
+#' ex1 <- group_apply(occdf = occdf, group = "cc", fun = nrow)
+#' # Temporal range of taxa per country with default arguments
+#' ex2 <- group_apply(occdf = occdf, group = "cc", fun = tax_range_time)
+#' # Temporal range of taxa per country with updated arguments
+#' ex3 <- group_apply(occdf = occdf,
+#'                    group = c("cc"),
+#'                    fun = tax_range_time,
+#'                    by = "LAD") # Order by LAD (default: "FAD")
+#' # Use multiple variables (number of occurrences per collection and formation)
+#' ex4 <- group_apply(occdf = occdf,
+#'                    group = c("collection_no", "formation"),
+#'                    fun = nrow)
+#' # Compute counts of occurrences per latitudinal bin
+#' # Set up lat bins
+#' bins <- lat_bins()
+#' # bin occurrences
+#' occdf <- bin_lat(occdf = tetrapods, bins = bins)
+#' # Calculate number of occurrences per bin
+#' ex5 <- group_apply(occdf = occdf, group = "lat_bin", fun = nrow)
+#' @export
+group_apply <- function(occdf, group, fun, ...) {
+
+  # Handle errors
+  if (!is.data.frame(occdf)) {
+    stop("`occdf` should be a dataframe")
+  }
+
+  if (!any(group %in% colnames(occdf))) {
+    stop("Supplied `group` is not a named column in `occdf`")
+  }
+
+  if (!is.function(fun)) {
+    stop("Supplied `fun` is not a function")
+  }
+
+  supp_args <- list(...)
+  indx <- which(!(names(supp_args) %in% names(formals(fun))))
+  if (length(indx) > 1) {
+    stop(paste(paste0("`", names(supp_args)[indx], "`", collapse = "/"),
+               "are not valid arguments for the specified function"))
+  } else if (length(indx) == 1) {
+    stop(paste0("`", names(supp_args)[indx], "`",
+                " is not a valid argument for the specified function"))
+  }
+
+  # Generate bin codes
+  bin_codes <- as.formula(paste0("~ ", paste(group, collapse = " + ")))
+  # Split dataframe
+  lst <- split(occdf, f = bin_codes, drop = TRUE, sep = "&1*^$0%")
+  # Group names
+  nme <- names(lst)
+  # Split into individual columns
+  nme_df <- do.call(rbind.data.frame, strsplit(nme, "&1*^$0%", fixed = TRUE))
+  colnames(nme_df) <- group
+  # Apply function
+  output_lst <- lapply(lst, fun, ...)
+  # Add groupings to output
+  output_df <- do.call(rbind.data.frame,
+                       lapply(seq_along(output_lst), FUN = function(i) {
+    df <- output_lst[[i]]
```

Comment:
```suggestion
    df <- data.frame(output_lst[[I]])
```
This should fix things! But some updating of column names might be needed.